README.md
# unzoner-dns
> [black.box Unzoner](https://unzoner.com) DNS server, inspired by [this](https://stackoverflow.com/a/33531753/1559300) post

This block implements a simple DNS server, which responds to `A` and `AAAA` dns queries,
using responses from [Unzoner API backend](https://github.com/belodetek/unzoner-api). The
format for the DNS query follows `{{alpha-2}}.{{dns_sub_domain}}.{{dns_domain}}`. ISO-3166
country codes are described [here](https://www.iban.com/country-codes) (e.g.)

```sh
$ dig +short CA.blackbox.unzoner.com
174.138.72.255
```

Delegate your `{{dns_sub_domain}}` under `{{dns_domain}}` to the endpoint running the DNS
server and ensure public access to `53/udp` port (e.g.)

```sh
$ dig +short NS blackbox.unzoner.com
ns1.blackbox.unzoner.com.
```

## usage
* ensure the API backend is deployed and configured
* add the latest [unzoner-dns](https://hub.balena.io/organizations/belodetek/blocks) block to your balenaCloud fleet composition (e.g. `amd64`)

```yml
version: '2.4'

services:
  unzoner-api:
    ...

  unzoner-dns:
    # https://www.balena.io/docs/learn/develop/blocks/#using-your-block-in-other-projects
    image: bh.cr/belodetek/unzoner-dns-amd64
    restart: unless-stopped
    ports:
      - "53:53/udp"
    # https://www.balena.io/docs/reference/supervisor/docker-compose/#labels
    labels:
      io.balena.update.strategy: download-then-kill
```

* set `API_SECRET` fleet environment variable to match API backend


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v0.0.3](https://github.com/belodetek/unzoner-dns/compare/v0.0.2...v0.0.3)

> 10 September 2022

- fix: return default records [`#3`](https://github.com/belodetek/unzoner-dns/pull/3)

#### [v0.0.2](https://github.com/belodetek/unzoner-dns/compare/v0.0.1...v0.0.2)

> 8 September 2022

- fix: init changelog [`#2`](https://github.com/belodetek/unzoner-dns/pull/2)
- ci: Update CHANGELOG and v0.0.2 asset(s) [`a7c757b`](https://github.com/belodetek/unzoner-dns/commit/a7c757b40dbe3a25888e2f7fe72e723f6ee5c77c)

#### v0.0.1

> 5 September 2022

- minor: create balenaBlock(s) [`#1`](https://github.com/belodetek/unzoner-dns/pull/1)
- Initial commit [`0d7d89f`](https://github.com/belodetek/unzoner-dns/commit/0d7d89ff3cdc24674ae325840bacd0b52792d539)
- ci: Update CHANGELOG and v0.0.1 asset(s) [`11330ae`](https://github.com/belodetek/unzoner-dns/commit/11330aebb37d3d4bd8fb503d73b5c4741403cccc)


src/utils.py
# -*- coding: utf-8 -*-

import json

from time import sleep
from requests import get
from functools import wraps
from subprocess import Popen, PIPE
from inspect import stack

try:
    from urllib import quote, unquote
except ImportError:
    from urllib.parse import quote, unquote

try:
    from httplib import (NOT_FOUND, OK)
except ImportError:
    from http.client import (NOT_FOUND, OK)

from config import *


def retry(ExceptionToCheck, tries=DEFAULT_TRIES, delay=DEFAULT_DELAY, backoff=DEFAULT_BACKOFF, cdata=None):
    '''Retry calling the decorated function using an exponential backoff.
    http://www.saltycrane.com/blog/2009/11/trying-out-retry-decorator-python/
    original from: http://wiki.python.org/moin/PythonDecoratorLibrary#Retry
    :param ExceptionToCheck: the exception to check. may be a tuple of
        exceptions to check
    :type ExceptionToCheck: Exception or tuple
    :param tries: number of times to try (not retry) before giving up
    :type tries: int
    :param delay: initial delay between retries in seconds
    :type delay: int
    :param backoff: backoff multiplier e.g. value of 2 will double the delay
        each retry
    :type backoff: int
    :param logger: logger to use. If None, print
    :type logger: logging.Logger instance
    '''
    def deco_retry(f):
        @wraps(f)
        def f_retry(*args, **kwargs):
            mtries, mdelay = tries, delay
            while mtries > 0:
                try:
                    return f(*args, **kwargs)
                except ExceptionToCheck as e:
                    print(
                        '{}, retrying in {} seconds (mtries={}): {}'.format(
                            repr(e),
                            mdelay,
                            mtries,
                            str(cdata)
                        )
                    )
                    sleep(mdelay)
                    mtries -= 1
                    mdelay *= backoff
            return f(*args, **kwargs)
        return f_retry  # true decorator
    return deco_retry


@retry(Exception, cdata='method={}'.format(stack()[0][3]))
def get_node(proto=4, country=TARGET_COUNTRY, client_ip='127.0.0.1'):
    country = quote(country)
    headers = {'X-Auth-Token': API_SECRET}
    try:
        res = get(
            '{}/api/v{}/node/{}/country/{}/geo/{}'.format(
                API_HOST,
                API_VERSION,
                proto,
                country,
                client_ip
            ),
            headers=headers
        )
        if DEBUG: print(res.status_code, res.content)
        if res.status_code in [NOT_FOUND]: return

        try:
            assert res.status_code in [OK]
            try:
                return json.loads(res.content)
            except:
                return res.content
        except:
            res = get(
                '{}/api/v{}/node/{}/country/{}'.format(
                    API_HOST,
                    API_VERSION,
                    proto,
                    country
                ),
                headers=headers
            )
            if DEBUG: print(res.status_code, res.content)
            if res.status_code in [NOT_FOUND]: return

            try:
                assert res.status_code in [OK]
                try:
                    return json.loads(res.content)
                except:
                    return res.content
            except:
                return
    except:
        raise AssertionError((res.status_code, res.content))


@retry(Exception, cdata='method={}'.format(stack()[0][3]))
def run_shell_cmd(cmd):
    shcmd = [unquote(el) for el in cmd.split(' ')]
    if DEBUG: print(shcmd)
    p = Popen(
        shcmd,
        stdin=PIPE,
        stdout=PIPE,
        stderr=PIPE,
        shell=False
    )
    output, err = p.communicate()
    return p.returncode, output, err


@retry(Exception, cdata='method={}'.format(stack()[0][3]))
def get_ipv(proto=4, host=MGMT_HOST):
    return run_shell_cmd('curl -%s %s' % (proto, host))[1]


@retry(Exception, cdata='method={}'.format(stack()[0][3]))
def get_countries():
    headers = {'X-Auth-Token': API_SECRET}
    res = get(
        '{}/api/v{}/countries/all'.format(
            API_HOST,
            API_VERSION
        ),
        headers=headers
    )
    if DEBUG: print(
        '{}: status_code={} content={}'.format(
            stack()[0][3],
            res.status_code,
            res.content
        )
    )
    if res.status_code not in [OK]:
        raise AssertionError((res.status_code, res.content))
    return res.content


src/dnsserver.py
#!/usr/bin/env python3

import sys
import json
import threading

from time import sleep, strftime, gmtime
from datetime import datetime
from dnslib.server import DNSServer
from dnslib import DNSRecord, CNAME, A,AAAA, RR, SOA, NS, QTYPE

from config import *
from utils import *

try:
	DEFAULT_A
except NameError:
	DEFAULT_A = None

try:
	DEFAULT_AAAA
except NameError:
	DEFAULT_AAAA = None

cache_ipv = {
	4: {
		"ip": get_ipv(),
		"ts": datetime.utcnow()
	},
	6: {
		"ip": get_ipv(6),
		"ts": datetime.utcnow()
	}
}

cache_node = {
	4: dict(), 6: dict()
}

class DomainName(str):
	def __getattr__(self, item):
		return DomainName(item + '.' + self)


class BlackboxResolver:
	def resolve(self, request, handler):
		if DEBUG: print(request, handler, handler.__dict__)
		domain = DomainName(DNS_DOMAIN)
		reply = request.reply()
		host = str(request.q.qname).lower()

		ttl = (datetime.utcnow() - cache_ipv[4]['ts']).total_seconds()
		if ttl > DNS_TTL:
			cache_ipv[4]['ts'] = datetime.utcnow()
			cache_ipv[4]['ip'] = get_ipv()
			if DEBUG: print('cache(miss): ip={} ttl={}'.format(
				cache_ipv[4]['ip'],
				ttl
			))
		ipv4 = cache_ipv[4]['ip']

		ttl = (datetime.utcnow() - cache_ipv[6]['ts']).total_seconds()
		if ttl > DNS_TTL:
			cache_ipv[6]['ts'] = datetime.utcnow()
			cache_ipv[6]['ip'] = get_ipv(6)
			if DEBUG: print('cache(miss): ip={} ttl={}'.format(
				cache_ipv[6]['ip'],
				ttl
			))
		ipv6 = cache_ipv[6]['ip']

		if ipv4 == ipv6: ipv6 = None
		t = int(strftime('%Y%m%d01', gmtime()))
		c_ip = handler.client_address[0]

		soa_record = SOA(
			mname=domain.ns1,
			rname=domain.admin,
			times=(
				t,
				60 * 60 * 1,
				60 * 60 * 3,
				60 * 60 * 24,
				60 * 60 * 1
			)
		)

		ns_record = NS(domain.ns1)
		records = [A(ipv4.decode()), soa_record, ns_record]
		if ipv6: records.append(AAAA(ipv6.decode()))

		records = {
			domain: records,
			domain.ns1: [A(ipv4.decode())],
			domain.admin: [CNAME(domain)],
		}

		print(
			'client={} c_ip={} qtype={} qname={} ipv4={} ipv6={}'.format(
				handler.client_address,
				c_ip,
				request.q.qtype,
				host,
				ipv4,
				ipv6
			)
		)

		if request.q.qtype in [6]: # SOA
			reply.add_answer(
				RR(
					rname=domain,
					rtype=QTYPE.SOA,
					rclass=1,
					ttl=DNS_TTL*5,
					rdata=soa_record
				)
			)
			return reply

		if request.q.qtype in [2]: # NS
			if str(request.q.qname) == domain:
				reply.add_answer(
					RR(
						rname=domain,
						rtype=QTYPE.NS,
						rclass=1,
						ttl=DNS_TTL*5,
						rdata=ns_record
					)
				)
			else:
				reply.add_answer(
					RR(
						rname=domain,
						rtype=QTYPE.SOA,
						rclass=1,
						ttl=DNS_TTL*5,
						rdata=soa_record
					)
				)
			return reply

		if str(request.q.qname) in [str(NS(domain.ns1)), domain]:
			if request.q.qtype in [1]: # A
				if ipv4: reply.add_answer(
					*RR.fromZone(
						'{} {} A {}'.format(
							host,
							DNS_TTL*5,
							ipv4
						)
					)
				)

			if request.q.qtype in [28]: # AAAA
				if ipv6: reply.add_answer(
					*RR.fromZone(
						'{} {} AAAA {}'.format(
							host,
							DNS_TTL*5,
							ipv6
						)
					)
				)
			return reply

		try:
			country = ISO_MAP[host.split('.')[0].upper()]
		except KeyError:
			if request.q.qtype in [1]: # A
				if DEFAULT_A:
					reply.add_answer(
						*RR.fromZone(
							'{} {} A {}'.format(
								host,
								0,
								DEFAULT_A
							)
						)
					)
				return reply
			elif request.q.qtype in [28]: # AAAA
				if DEFAULT_AAAA:
					reply.add_answer(
						*RR.fromZone(
							'{} {} AAAA {}'.format(
								host,
								0,
								DEFAULT_AAAA
							)
						)
					)
			else:
				return reply

		if request.q.qtype in [1]: # A
			af = 4
			try:
				try:
					ttl = (datetime.utcnow() - cache_node[af][country]['ts']).total_seconds()
				except:
					cache_node[af][country] = dict()
					cache_node[af][country]['ts'] = datetime.utcnow()
					cache_node[af][country]['ip'] = None
					ttl = DNS_TTL + 1

				if ttl > DNS_TTL:
					cache_node[af][country]['ts'] = datetime.utcnow()
					cache_node[af][country]['ip'] = get_node(
						client_ip=c_ip,
						country=country
					)
					if DEBUG: print('cache(miss): ip={} ttl={} country={}'.format(
						cache_node[af][country]['ip'],
						ttl,
						country
					))
				ip = cache_node[af][country]['ip'].decode()
				if DEBUG: print('cache={} ip={}'.format(cache_node, ip))
				assert ip
			except:
				if DEFAULT_A:
					reply.add_answer(
						*RR.fromZone(
							'{} {} A {}'.format(
								host,
								0,
								DEFAULT_A
							)
						)
					)
				return reply

			reply.add_answer(
				*RR.fromZone(
					'{} {} A {}'.format(
						host,
						DNS_TTL,
						ip
					)
				)
			)

		elif request.q.qtype in [28]: # AAAA
			af = 6
			try:
				try:
					ttl = (datetime.utcnow() - cache_node[af][country]['ts']).total_seconds()
				except:
					cache_node[af][country] = dict()
					cache_node[af][country]['ts'] = datetime.utcnow()
					cache_node[af][country]['ip'] = None
					ttl = DNS_TTL + 1

				if ttl > DNS_TTL:
					cache_node[af][country]['ts'] = datetime.utcnow()
					cache_node[af][country]['ip'] = get_node(
						proto=af,
						client_ip=c_ip,
						country=country
					)
					if DEBUG: print('cache(miss): ip={} ttl={} country={}'.format(
						cache_node[af][country]['ip'],
						ttl,
						country
					))
				ip = cache_node[af][country]['ip'].decode()
				if DEBUG: print('cache={} ip={}'.format(cache_node, ip))
				assert ip
			except:
				if DEFAULT_AAAA:
					reply.add_answer(
						*RR.fromZone(
							'{} {} AAAA {}'.format(
								host,
								0,
								DEFAULT_AAAA
							)
						)
					)
				return reply

			reply.add_answer(
				*RR.fromZone(
					'{} {} AAAA {}'.format(
						host,
						DNS_TTL,
						ip
					)
				)
			)

		else: print(request.q.qtype, country, host)
		return reply


if __name__ == '__main__':
	try:
		countries = json.loads(get_countries())
		ISO_MAP = {v: k for k, v in countries.items()}
	except AssertionError:
		sys.exit(1)

	resolver = BlackboxResolver()

	server = DNSServer(
		resolver,
		port=PORT,
		address=LISTEN_ADDR,
		logger=None,
		tcp=False
	)

	thread = threading.Thread(target=server.start_thread())
	thread.daemon = True
	thread.start()

	print(
		'server running in thread: {} ({}:{})'.format(
			thread.name,
			LISTEN_ADDR,
			PORT
		)
	)

	try:
		while True:
			sleep(1)
			sys.stderr.flush()
			sys.stdout.flush()

	except KeyboardInterrupt:
		pass


src/config.py
# -*- coding: utf-8 -*-

import os

DEFAULT_A = os.getenv('DEFAULT_A', '174.138.72.255')
DEFAULT_AAAA = os.getenv('DEFAULT_AAAA', '2604:a880:800:10::128:6001')
API_VERSION = os.getenv('API_VERSION', '1.0')
TARGET_COUNTRY = os.getenv('TARGET_COUNTRY', 'United States')
API_SECRET = os.getenv('API_SECRET', None)
API_HOST = os.getenv('API_HOST', 'https://api.unzoner.com')
DEBUG = int(os.getenv('DEBUG', 1))
DEFAULT_TRIES = int(os.getenv('DEFAULT_TRIES', 3))
DEFAULT_DELAY = int(os.getenv('DEFAULT_DELAY', 2))
DEFAULT_BACKOFF = int(os.getenv('DEFAULT_BACKOFF', 2))
DNS_DOMAIN = os.getenv('DNS_DOMAIN', 'blackbox.unzoner.com.')
LISTEN_ADDR = os.getenv('LISTEN_ADDR', '0.0.0.0')
PORT = int(os.getenv('PORT', 53))
MGMT_HOST = os.getenv('MGMT_HOST', 'mgmt.unzoner.com')
DNS_TTL = int(os.getenv('DNS_TTL', 60)) # seconds
